planner: update the doc of SPM (#13883)
pingcap/docs: optimizer-hints.md

| └─IndexReader_22(Probe)     | 9990.00  | root      |                      | index:IndexFullScan_21                       |
|   └─IndexFullScan_21        | 9990.00  | cop[tikv] | table:t1, index:k(k) | keep order:false, stats:pseudo               |
+-----------------------------+----------+-----------+----------------------+----------------------------------------------+
5 rows in set, 1 warning (0.00 sec)
```

In the preceding statements, the collations of `t1.k` and `t2.k` are different (`utf8mb4_general_ci` and `utf8mb4_bin` respectively), which makes the `INL_JOIN` or `TIDB_INLJ` hint cannot take effect.

```sql
SHOW WARNINGS;
+---------+------+----------------------------------------------------------------------------+
| Level   | Code | Message                                                                    |
+---------+------+----------------------------------------------------------------------------+
| Warning | 1815 | Optimizer Hint /*+ INL_JOIN(t1) */ or /*+ TIDB_INLJ(t1) */ is inapplicable |
+---------+------+----------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

> ### `INL_JOIN` Hint does not take effect because of join order

Reviewer: ```suggestion
### `INL_JOIN` hint does not take effect because of join order
```

> The meaning of `INL_JOIN(t1, t2)` or `TIDB_INLJ(t1, t2)` is to use `IndexJoin` for `t1` and `t2` individually, instead of using an `IndexJoin` to join them together directly. For example:

Reviewer: ```suggestion
The [`INL_JOIN(t1, t2)`](#inl_joint1_name--tl_name-) or `TIDB_INLJ(t1, t2)` hint uses the `IndexJoin` operator when joining `t1` and `t2` with other tables, instead of using an `IndexJoin` operator to directly join them together. For example:
```

| │   └─TableRangeScan_30         | 10000.00 | cop[tikv] | table:t1      | range: decided by [test.t2.id], keep order:false, stats:pseudo                                                                                                          |
| └─TableReader_13(Probe)         | 12500.00 | root      |               | data:TableRangeScan_12                                                                                                                                                  |
|   └─TableRangeScan_12           | 12500.00 | cop[tikv] | table:t3      | range: decided by [test.t2.id test.t1.id], keep order:false, stats:pseudo                                                                                               |
+---------------------------------+----------+-----------+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

> In the case above, `t1` and `t3` are not joined togather by an `IndexJoin` directly.

Reviewer: ```suggestion
In the preceding example, `t1` and `t3` are not directly joined together by an `IndexJoin`.
```

> If you want to use an `IndexJoin` to join `t1` and `t3` directly, use `LEADING(t1, t3)` Hint first to specify the join order, and then use `INL_JOIN` to specify the join algorithm. For example:

Reviewer: ```suggestion
To perform a direct `IndexJoin` between `t1` and `t3`, you can first use [`LEADING(t1, t3)` hint](#leadingt1_name--tl_name-) to specify the join order of `t1` and `t3`, and then use the `INL_JOIN` hint to specify the join algorithm. For example:
```
